File: Yandex_training/Algorithm_training 9.0/0_tests/2.py
import sys
from itertools import accumulate
input = sys.stdin.readline

def solve():
    n = int(input())
    a = list(map(int, input().split()))
    
    if n == 1:
        sys.stdout.write("1\n")
        return
    
    # prefix sums
    prefix = list(accumulate(a))  # prefix[i] = sum(a[0..i])
    
    # last_break: last j (1-indexed compare) where sum(a[0..j-1]) <= a[j]
    # i.e., prefix[j-1] <= a[j]
    last_break = 0
    for j in range(1, n):
        if prefix[j-1] <= a[j]:
            last_break = j
    
    # first index where a[i] > a[0]
    first_can_start = n
    # a is sorted, use binary search equivalent
    lo, hi = 0, n
    while lo < hi:
        mid = (lo + hi) // 2
        if a[mid] > a[0]:
            hi = mid
        else:
            lo = mid + 1
    first_can_start = lo
    
    first_winner = max(last_break, first_can_start)
    
    out = ['0'] * n
    for i in range(first_winner, n):
        out[i] = '1'
    
    sys.stdout.write(' '.join(out) + '\n')

solve()

File: Yandex_training/Algorithm_training 9.0/0_tests/2.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    vector<long long> a(n);
    for (int i = 0; i < n; i++) cin >> a[i];

    if (n == 1) {
        cout << 1 << "\n";
        return 0;
    }

    // prefix[i] = sum(a[0..i-1]), prefix[0] = 0
    vector<long long> prefix(n + 1, 0);
    for (int i = 0; i < n; i++)
        prefix[i + 1] = prefix[i] + a[i];

    // last_break: last j where sum(a[0..j-1]) <= a[j]
    int last_break = 0;
    for (int j = 1; j < n; j++)
        if (prefix[j] <= a[j])
            last_break = j;

    // first index where a[i] > a[0] (binary search)
    int first_can_start = (int)(upper_bound(a.begin(), a.end(), a[0]) - a.begin());

    int first_winner = max(last_break, first_can_start);

    for (int i = 0; i < n; i++)
        cout << (i >= first_winner ? 1 : 0) << " \n"[i == n - 1];

    return 0;
}